### PR TITLE
Optimistic locking (or not) during delete.

### PR DESCRIPTION
**Pull request by meistermeier**
Currently marked as draft because the reactive part will be done after we agree on this.

### CODE REVIEW

File: src/test/java/org/springframework/data/neo4j/integration/reactive/ReactiveOptimisticLockingIT.java [hunk 1]
```diff
@@ -204,6 +204,80 @@ void shouldFailIncrementVersionsForMultipleSaveForAssignedId(
...
+				.verifyComplete();
+
+	}
+
+	@Test
+	void shouldNotFailOnDeleteByIdWithAnyVersion(@Autowired VersionedThingWithAssignedIdRepository repository) {
+		try (Session session = driver.session()) {
+			session.writeTransaction(tx -> {
+				tx.run("CREATE (v:VersionedThingWithAssignedId {id:1, myVersion:3})").consume();
+				return null;
```
Comment (michael-simons) [inline src/test/java/org/springframework/data/neo4j/integration/reactive/ReactiveOptimisticLockingIT.java:248]:
```suggestion
				return tx.run("CREATE (v:VersionedThingWithAssignedId {id:1, myVersion:3})").consume();
```

File: src/test/java/org/springframework/data/neo4j/integration/reactive/ReactiveOptimisticLockingIT.java [hunk 2]
```diff
@@ -204,6 +204,80 @@ void shouldFailIncrementVersionsForMultipleSaveForAssignedId(
...
+			assertThat(count).isEqualTo(0);
+		}
+	}
+
+	@Test
+	void shouldFailOnDeleteByEntityWithWrongVersion(@Autowired VersionedThingWithAssignedIdRepository repository) {
+		try (Session session = driver.session()) {
+			session.writeTransaction(tx -> {
+				tx.run("CREATE (v:VersionedThingWithAssignedId {id:1, myVersion:2})").consume();
+				return null;
```
Comment (michael-simons) [inline src/test/java/org/springframework/data/neo4j/integration/reactive/ReactiveOptimisticLockingIT.java:269]:
```suggestion
				return tx.run("CREATE (v:VersionedThingWithAssignedId {id:1, myVersion:2})").consume();
```

**Review by michael-simons — CHANGES_REQUESTED** [review]
Only minor change request: I don't like the return null ;)
One question (regarding the relationships properties)

File: etc/adr/adr-007.adoc [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+There should be a check against the version of the previously loaded entity.
+As a consequent throw a `OptimisticLockingFailureException` if the versions have a mismatch on delete.
+
+Non-existing version parameters during `delete(entity)`:
+Fall back to `OR <versionField> IS NULL` to provide a relaxed mode.
+Otherwise, we would initialize the version actively to just delete the node.
+
+=== Consequences
+
+Users will have to refactor the existing `@RelationshipProperties` based relationships.
```
[the remark this answers is not in the document]
Comment (meistermeier) [reply etc/adr/adr-007.adoc:29]: Just because my copy and paste skills are low..

**Comment by meistermeier** [comment]
Fixed with e7989df04f92fe14feaa8d71fb88f3558717e602 on 6.1 and 881e741d123a67ba2e672d6b686afac4772bb6ef on 6.0

**Pull request closed without merging**